### PR TITLE
Add a GitHub action/CLI tool for analyzing vulnerabilities.

### PR DESCRIPTION
**Pull request by oliverchang**
This action detects changed vulnerability files and performs analysis on them.
Analysis includes detecting impacted versions from the given input ranges.

Also make some optimizations to make analysis work for large repos (like
the Linux kernel):
- Optimized implementation of detecting tags which contain a commit.
- Add option to disable detecting cherry picks.

This tool can also be run as a CLI standalone tool:

```bash
$ python -m osv.analyze_tool ...
```

### CODE REVIEW

**Review by inferno-chromium — APPROVED** [review]

File: lib/osv/sources.py [hunk 1]
```diff
@@ -84,7 +109,37 @@ def vulnerability_to_yaml(vulnerability, output_path):
...
+  """Update a vulnerability file on disk."""
+  if os.path.exists(output_path):
+    data = _parse_vulnerability_dict(output_path)
+  else:
+    data = {}
+    if key_path:
+      cur = data
+      for component in key_path.split('.'):
+        cur[component] = {}
+        cur = cur[component]
```
[the remark this answers is not in the document]
Comment (oliverchang) [reply lib/osv/sources.py:138]: done!

File: lib/osv/analyze_tool/__init__.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+from .. import sources
+
+
+def main():
+  logging.basicConfig(level=logging.INFO)
+  parser = argparse.ArgumentParser(description='Analyze')
+  parser.add_argument(
+      '--analyze_git',
+      help='Analyze git ranges',
+      choices=['true', 'false'],
```
[the remark this answers is not in the document]
Comment (oliverchang) [reply lib/osv/analyze_tool/__init__.py:35]: This makes it easier to use from e.g. a github actions config without needing to have if statements.

this style is also supported by e.g. gflags (or abseil). 

We can use something like https://abseil.io/docs/python/guides/flags to make this accept both `flag=true,false` or `--flag`, but it felt a bit overkill in this case.

File: lib/osv/analyze_tool/__init__.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  args = parser.parse_args()
+
+  analyze_git = args.analyze_git == 'true'
+  detect_cherrypicks = args.detect_cherrypicks == 'true'
+
+  if args.pr_base:
+    paths = subprocess.check_output(
+        ['git', 'diff', '--name-only', args.pr_base]).decode().splitlines()
+  else:
+    paths = args.paths
```
[the remark this answers is not in the document]
Comment (oliverchang) [reply lib/osv/analyze_tool/__init__.py:64]: can't be None, but it could be empty. added a print in that case.

**Review by oliverchang — COMMENTED** [review]
thanks!

**Pull request merged**